### PR TITLE
Fix Issue 21489 - Duplicated template instantiation using mixin insid…

### PR DESCRIPTION
**Pull request by BorisCarvajal**
…e alias declaration

### CODE REVIEW

File: src/dmd/typesem.d [hunk 1]
```diff
@@ -1969,22 +1969,16 @@ extern(C++) Type typeSemantic(Type type, const ref Loc loc, Scope* sc)
...
+
+        Expression e;
+        Type t;
+        Dsymbol s;
+        mtype.resolve(loc, sc, e, t, s);
+
+        if (t && t.ty != Terror)
+            return t;
+
+        .error(mtype.loc, "`mixin(%s)` does not give a valid type", mtype.obj.toChars);
```
Comment (RazvanN7) [inline src/dmd/typesem.d:1981]: This line is not covered by any tests

**Review by RazvanN7 — CHANGES_REQUESTED** [review]
Thank you for the contribution! However, I do have some requests:

- All code lines need to be covered by tests.
- Can you please explain in a few words what was the problem with the code and how your solution fixes it?

File: src/dmd/typesem.d (hunk 1 above)
Comment (BorisCarvajal) [reply src/dmd/typesem.d:1981]: test added

**Comment by BorisCarvajal** [comment]
>     * Can you please explain in a few words what was the problem with the code and how your solution fixes it?

The code inside `mixin` is resolved two times when creating an alias because `TypeMixin` doesn't cache the semantic evaluation and aliasSemantic() indirectly calls it twice.
That leads to for example, creating two instances of a template if using this kind of `mixin`.

**Comment by 12345swordy** [comment]
Comment your code on why are you doing this, it hard for newcomers to understand the internals of the compiler already as it.

**Comment by BorisCarvajal** [comment]
> Comment your code on why are you doing this, it hard for newcomers to understand the internals of the compiler already as it.

Look at `src/dmd/mtype.d` changes:
```
RootObject obj; // cached result of semantic analysis.
```
It's easy to follow from that.

**Comment by 12345swordy** [comment]
> 
> 
> > Comment your code on why are you doing this, it hard for newcomers to understand the internals of the compiler already as it.
> 
> Look at `src/dmd/mtype.d` changes:
> 
> ```
> RootObject obj; // cached result of semantic analysis.
> ```
> 
> It's easy to follow from that.

No it is not. There are multiple files that are currently modified by this.

File: src/dmd/typesem.d [hunk 2]
```diff
@@ -3079,8 +3073,16 @@ void resolve(Type mt, const ref Loc loc, Scope* sc, out Expression pe, out Type
...
     {
-        auto o = mt.compileTypeMixin(loc, sc);
+        RootObject o = mt.obj;
+        if (o)
+        {
+            pe = o.isExpression();
+            pt = o.isType();
+            ps = o.isDsymbol();
+            return;
+        }
```
Comment (12345swordy) [inline src/dmd/typesem.d:3083]: Comment on why are you checking this.

File: src/dmd/typesem.d [hunk 3]
```diff
@@ -3091,12 +3093,14 @@ void resolve(Type mt, const ref Loc loc, Scope* sc, out Expression pe, out Type
...
             if (auto et = e.isTypeExp())
-                return returnType(et.type.addMod(mt.mod));
+                returnType(et.type.addMod(mt.mod));
             else
                 returnExp(e);
         }
         else
             returnError();
+
+        mt.obj = pe ? pe : (pt ? pt : ps);
```
Comment (12345swordy) [inline src/dmd/typesem.d:3103]: Same thing here. Comment on why please.

**Review by 12345swordy — CHANGES_REQUESTED** [review]
See my comments

File: src/dmd/typesem.d (hunk 2 above)
Comment (BorisCarvajal) [reply src/dmd/typesem.d:3083]: Remember: `RootObject obj; // cached result of semantic analysis.`

I'm checking if the object is already cached, if so just return it in the corresponding variable.

File: src/dmd/typesem.d (hunk 3 above)
Comment (BorisCarvajal) [reply src/dmd/typesem.d:3103]: Here the opposite. I'm saving the semantic result.

**Review by RazvanN7 — APPROVED** [review]

**Pull request merged**